networks:
  traefik-net:
    external: true

services:
  traefik:
    image: traefik:v3.5.0
    container_name: traefik
    restart: always
    command: >
      sh -c "touch /letsencrypt/acme_tls.json \
      && chmod 600 /letsencrypt/acme_tls.json \
      && touch /letsencrypt/acme_dns.json \
      && chmod 600 /letsencrypt/acme_dns.json \
      && touch /letsencrypt/acme_dns_account.json \
      && chmod 600 /letsencrypt/acme_dns_account.json \
      && touch /letsencrypt/acme_http.json \
      && chmod 600 /letsencrypt/acme_http.json \
      && traefik \
      --log.level=INFO
      --api.dashboard=true
      --api=true
      --api.insecure=true
      --providers.docker=true
      --providers.docker.exposedbydefault=false
      --entrypoints.web.address=:80
      --entrypoints.websecure.address=:443
      --entrypoints.metrics.address=:8082
      --entrypoints.web.http.redirections.entrypoint.to=websecure
      --entrypoints.web.http.redirections.entrypoint.scheme=https
      --entrypoints.web.http.redirections.entrypoint.permanent=true
      --metrics.prometheus=true
      --metrics.prometheus.entrypoint=metrics
      --metrics.prometheus.addEntryPointsLabels=true
      --metrics.prometheus.addRoutersLabels=true
      --metrics.prometheus.addServicesLabels=true
      --certificatesresolvers.tls-resolver.acme.tlschallenge=true
      --certificatesresolvers.tls-resolver.acme.email=desarrollo@indaws.es
      --certificatesresolvers.tls-resolver.acme.storage=/letsencrypt/acme_tls.json
      --certificatesresolvers.dns-resolver.acme.dnschallenge=true
      --certificatesresolvers.dns-resolver.acme.dnschallenge.provider=acme-dns
      --certificatesresolvers.dns-resolver.acme.email=desarrollo@indaws.es
      --certificatesresolvers.dns-resolver.acme.dnschallenge.resolvers=8.8.8.8:53,1.1.1.1:53
      --certificatesresolvers.dns-resolver.acme.storage=/letsencrypt/acme_dns.json
      --certificatesresolvers.http-resolver.acme.httpchallenge=true
      --certificatesresolvers.http-resolver.acme.httpchallenge.entrypoint=web
      --certificatesresolvers.http-resolver.acme.email=desarrollo@indaws.es
      --certificatesresolvers.http-resolver.acme.storage=/letsencrypt/acme_http.json"
    environment:
      - ACME_DNS_API_BASE=https://acme.indaws.cloud
      - ACME_DNS_STORAGE_PATH=/letsencrypt/acme_dns_account.json
    ports:
      - "80:80"
      - "443:443"
      - "8082"
    networks:
      - traefik-net
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./letsencrypt:/letsencrypt
    labels:
      - "traefik.enable=true"

      # middlewares
      - "traefik.http.middlewares.gzip.compress=true"
      - "traefik.http.middlewares.incloud-ip-whitelist.ipallowlist.sourcerange=194.30.4.97/32, 194.30.4.101/24, 172.16.100.1/24, 10.10.10.1/24"
    logging:
      driver: json-file
      options:
        max-size: 10m

  # error page
  errorpage:
    image: nginx:alpine
    container_name: traefik-errorpage
    restart: always
    volumes:
      - ./errorpage:/usr/share/nginx/html:ro
      - ./nginx-error.conf:/etc/nginx/conf.d/default.conf:ro
    networks:
      - traefik-net
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik-net"

      # maintenance page
      - "traefik.http.services.maintenance-service.loadbalancer.server.port=80"
      - "traefik.http.services.maintenance-service.loadbalancer.server.scheme=http"
      - "traefik.http.services.maintenance-service.loadbalancer.passhostheader=false"

      # known domains - priority very low
      - "traefik.http.routers.errorpage-fallback.rule=PathPrefix(`/`)"
      - "traefik.http.routers.errorpage-fallback.entrypoints=websecure"
      - "traefik.http.routers.errorpage-fallback.tls=true"
      - "traefik.http.routers.errorpage-fallback.service=maintenance-service"
      - "traefik.http.routers.errorpage-fallback.priority=-1"

      # 403 middleware
      - "traefik.http.middlewares.add-forbidden-header.headers.customrequestheaders.X-Forbidden-Access=true"
    logging:
      driver: json-file
      options:
        max-size: 10m
